### ISSUE TITLE
PyLance/PyRight resolving to site-packages instead of local workspace files

### ISSUE THREAD

**Issue by cspotcode** [microsoft/pyright#5719]
**Describe the bug**

I am editing a python library locally.  Another version of this library is also installed via pip.  "Go to Definition" navigates to the pip-installed version of the library instead of the local one.  As far as I can tell, I've configured everything correctly.

For imports that can be found *both* locally and in my pip site-packages, PyLance/PyRight is resolving to files in pip site-packages.  For imports that *only* exist locally, it is resolving to the local file.

I am expecting it to resolve to the local file in *both* situations, and as far as I can tell, this should happen with the default pyright configuration.

**Code or Screenshots**

I've opened an empty directory in VSCode and populated it with the following:
```
pyproject.toml
arcade/__init__.py
arcade/foo.py
arcade/sprite.py
```
```toml
# pyproject.toml
[tool.pyright]
# nothing here
```
```python
# arcade/__init__.py
from arcade.foo import foo
from arcade.sprite import foo as foo2
```
```python
# arcade/foo.py
def foo(): pass
```
```python
# arcade/sprite.py
def bar(): pass
```

I am not using a virtual env.  I have installed the `arcade` library from pip, though I believe this repro should work with any installed library.

"Go to Declaration" for `arcade.foo` navigates to `foo.py`, since the pip installation of arcade does not have a `foo` submodule.  "Go to Declaration" for `arcade.sprite` navigates to `C:\Users\cspot\scoop\persist\python\Lib\site-packages\arcade\sprite\__init__.py`, when I believe it should navigate to the workspace's `sprite.py`.

**VS Code extension or command-line**

VSCode PyLance v2023.8.20

**PyLance logs**

Here is the start of my PyLance log, which shows the search paths it is using.  I have turned on "Trace" logging level.  I can share the full log if it'll help.

```
2023-08-14 16:48:35.336 [info] (2384) WorkspaceFactory 0 clear
2023-08-14 16:48:35.338 [info] (Client) Telemetry is disabled
2023-08-14 16:48:35.338 [info] (Client) Experiments are disabled, only manually opted experiments are active.
2023-08-14 16:48:36.555 [info] [Info  - 4:48:36 PM] (5392) Pylance language server 2023.8.20 (pyright fcee52c7) starting
2023-08-14 16:48:36.556 [info] [Info  - 4:48:36 PM] (5392) Server root directory: c:\Users\cspot\.vscode\extensions\ms-python.vscode-pylance-2023.8.20\dist
2023-08-14 16:48:36.561 [info] [Info  - 4:48:36 PM] (5392) Starting service instance "pyright-resolution-bug"
2023-08-14 16:48:36.582 [info] (5392) No configuration file found.
2023-08-14 16:48:36.582 [info] (5392) pyproject.toml file found at c:\Users\cspot\Documents\dev\pyright-resolution-bug.
2023-08-14 16:48:36.582 [info] [Info  - 4:48:36 PM] (5392) Setting pythonPath for service "pyright-resolution-bug": "C:\Users\cspot\scoop\apps\python\current\python.exe"
2023-08-14 16:48:36.586 [info] [Info  - 4:48:36 PM] (5392) Loading pyproject.toml file at c:\Users\cspot\Documents\dev\pyright-resolution-bug\pyproject.toml
2023-08-14 16:48:36.683 [info] [Info  - 4:48:36 PM] (5392) Assuming Python version 3.11
2023-08-14 16:48:36.684 [info] (5392) Assuming Python platform Windows
2023-08-14 16:48:36.684 [info] [Info  - 4:48:36 PM] (5392) No include entries specified; assuming c:\Users\cspot\Documents\dev\pyright-resolution-bug
2023-08-14 16:48:36.684 [info] [Info  - 4:48:36 PM] (5392) Auto-excluding **/node_modules
2023-08-14 16:48:36.684 [info] [Info  - 4:48:36 PM] (5392) Auto-excluding **/__pycache__
2023-08-14 16:48:36.684 [info] [Info  - 4:48:36 PM] (5392) Auto-excluding **/.*
2023-08-14 16:48:36.684 [info] [Warn  - 4:48:36 PM] (5392) stubPath c:\Users\cspot\Documents\dev\pyright-resolution-bug\typings is not a valid directory.
2023-08-14 16:48:36.803 [info] [Info  - 4:48:36 PM] (5392) Search paths for c:\Users\cspot\Documents\dev\pyright-resolution-bug
2023-08-14 16:48:36.803 [info] [Info  - 4:48:36 PM] (5392)   c:\Users\cspot\.vscode\extensions\ms-python.vscode-pylance-2023.8.20\dist\typeshed-fallback\stdlib
2023-08-14 16:48:36.804 [info] [Info  - 4:48:36 PM] (5392)   c:\Users\cspot\Documents\dev\pyright-resolution-bug
2023-08-14 16:48:36.804 [info] [Info  - 4:48:36 PM] (5392)   c:\Users\cspot\Documents\dev\pyright-resolution-bug\src
2023-08-14 16:48:36.804 [info] [Info  - 4:48:36 PM] (5392)   c:\Users\cspot\Documents\dev\pyright-resolution-bug\typings
2023-08-14 16:48:36.804 [info] [Info  - 4:48:36 PM] (5392)   c:\Users\cspot\.vscode\extensions\ms-python.vscode-pylance-2023.8.20\dist\typeshed-fallback\stubs\...
2023-08-14 16:48:36.805 [info] [Info  - 4:48:36 PM] (5392)   c:\Users\cspot\.vscode\extensions\ms-python.vscode-pylance-2023.8.20\dist\bundled\stubs
2023-08-14 16:48:36.805 [info] [Info  - 4:48:36 PM] (5392)   c:\Users\cspot\scoop\apps\python\3.11.2\DLLs
2023-08-14 16:48:36.805 [info] [Info  - 4:48:36 PM] (5392)   c:\Users\cspot\scoop\apps\python\3.11.2\Lib
2023-08-14 16:48:36.805 [info] [Info  - 4:48:36 PM] (5392)   c:\Users\cspot\scoop\apps\python\3.11.2
2023-08-14 16:48:36.806 [info] [Info  - 4:48:36 PM] (5392)   c:\Users\cspot\scoop\persist\python\Lib\site-packages
2023-08-14 16:48:36.806 [info] [Info  - 4:48:36 PM] (5392)   c:\Users\cspot\scoop\persist\python\Lib\site-packages\win32
2023-08-14 16:48:36.806 [info] [Info  - 4:48:36 PM] (5392)   c:\Users\cspot\scoop\persist\python\Lib\site-packages\win32\lib
2023-08-14 16:48:36.807 [info] [Info  - 4:48:36 PM] (5392)   c:\Users\cspot\scoop\persist\python\Lib\site-packages\pythonwin
2023-08-14 16:48:36.807 [info] [Info  - 4:48:36 PM] (5392) Adding fs watcher for library directories:
```

**Comment by cspotcode** [comment]
I tried to bisect pyright versions, to see when this behavior started.

Looks like it started in 1.1.308

https://github.com/microsoft/pyright/compare/1.1.307...1.1.308

I see that a resolution bugfix was made for 1.1.308, but I don't know if it's related at all.
https://github.com/microsoft/pyright/commit/43a6a8fae55c5b40ee040608271f9420a353d1a5

**Comment by cspotcode** [comment]
I forgot to mention, when bisecting to discover when the issue started, I used pyright CLI, not pylance.  `from sprite import bar` fails with a type error when pyright resolves incorrectly, because it resolves to the pip-installed `arcade.sprite` which does not export `bar`.

**Comment by erictraut** [comment]
Thanks for the bug report. Pyright is preferring the module in the site-packages directory because the package is marked "py.typed" but the local one isn't. I agree that this behavior isn't correct, so I've updated the logic accordingly. This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.323, which I just published. It will also be included in a future release of pylance.
